src/natcap/invest/forest_carbon_edge_effect.py

```diff
@@ -804,59 +808,60 @@ def _calculate_tropical_forest_edge_carbon_map(
             row_coords[valid_edge_distance_mask].ravel(),
             col_coords[valid_edge_distance_mask].ravel()))
         # note, the 'n_jobs' parameter was introduced in SciPy 0.16.0
-        distances, indexes = kd_tree.query(
+        # for each forest point x, for each of its k nearest neighbors
+        distances, indexes = kd_tree.query( # distances: (x, k) indexes: (x, k)
             coord_points, k=n_nearest_model_points,
             distance_upper_bound=DISTANCE_UPPER_BOUND, n_jobs=-1)
 
         if n_nearest_model_points == 1:
             distances = distances.reshape(distances.shape[0], 1)
             indexes = indexes.reshape(indexes.shape[0], 1)
 
-        # the 3 is for the 3 thetas in the carbon model
+        # 3 is for the 3 thetas in the carbon model. thetas shape: (x, k, 3)
         thetas = numpy.zeros((indexes.shape[0], indexes.shape[1], 3))
         valid_index_mask = (indexes != kd_tree.n)
         thetas[valid_index_mask] = theta_model_parameters[
             indexes[valid_index_mask]]
 
-        # the 3 is for the 3 models (asym, exp, linear)
-        biomass_model = numpy.zeros(
-            (indexes.shape[0], indexes.shape[1], 3))
         # reshape to an N,nearest_points so we can multiply by thetas
         valid_edge_distances_km = numpy.repeat(
             edge_distance_block[valid_edge_distance_mask] * cell_size_km,
             n_nearest_model_points).reshape(-1, n_nearest_model_points)
 
-        # asymptotic model
+        # For each forest pixel x, for each of its k nearest neighbors, the 
+        # chosen regression method (1, 2, or 3). model_index shape: (x, k)
+        model_index = numpy.empty(indexes.shape, dtype=numpy.int8) 
+        model_index[:] = 0
+        model_index[valid_index_mask] = ( 
+            method_model_parameter[indexes[valid_index_mask]]) 
+
+        # biomass shape: (x, k)
+        biomass = numpy.empty((indexes.shape[0], indexes.shape[1])) 
+        biomass[:] = 0
```

Comment:
Same, here, one step, but I think it's also a good idea to name the type if you know what it's going to be. I bet it'll be a float 32 so `biomass = numpy.zeros((indexes.shape[0], indexes.shape[1], dtype=numpy.float32)) `